### DISCUSSION TITLE
Galagino on TTGO T4

### DISCUSSION THREAD

**Discussion by C64fan** [harbaum/galagino#15] [General]
Hi,
progressing on the project.
All the python scripts succeeded on the roms but with compile of the project the galaga_starseed.h is throwing struct errors.
It was a bad sample_boom file, resolved with newly converted file.
Thank you

**Comment by harbaum** [comment]
What are those errors?

**Comment by harbaum** [comment]
I just did a build from scratch and it worked as it should. So there's something different in your setup.

My assumption is that one of the files was not created properly. It seems the first line if `galaga_starseed.h` is not parsed properly and thus the following second line in `galaga_starseed` fails. 

My guess is that the file `galaga_sample_boom.h` is broken as it's included right before `galaga_starseed.h` . Can you have a look at that file? It shouls start with `const unsigned char galaga_sample_boom[] = {` and end with `};` in the last line.

BTW: Where and how did you run the python scripts? Under Linux or Windows?

If everything else fails, please zip your entire galagino directory containing the sketch and send it to me via email, so i can have a look and try to figure out which file differs and why.

**Comment by C64fan** [comment]
Hi Till,
Little progress, the project compliles but the display stays dark/gray.
Are you happy with sharing your TFT_eSPI User_Setup.h and User_Setup_Select.h config files ?
Greets,

**Comment by harbaum** [comment]
I am not using any TFT library and thus no config files like that.

If you apply power and have the LED pin connected correctly then the ili9341 based display will light up bright white. If you have VCC (3.3V), GND and LED connected correctly then the screen should be white.

If that doesn't work and you are using a different setup than mine then you need to figure out how to configure Galagino for that specific board. You need to control the Backlight, drive the displays reset line correctly and finally setup the correct SPI pins.

I don't own one of the boards you use and can't easily help you with that.

**Comment by C64fan** [comment]
Hi Till,
your answer was magic, many thanks.
Managed to resolve the pinning, all working ok.
#define TFT_CS  27, #define TFT_DC  32, #define TFT_RST 5, #define TFT_BL  4 for LilyGO TTGO T4 V1.3 ESP32.

**Comment by harbaum** [comment]
Great. How about audio and the controls? How did you solve that?

I'll add your findings to the repository somehow so other TTGO T4 users can find tem.

**Comment by C64fan** [comment]
Hi Till,
making the cabinet out of thin wood with smd audio amplifier, it will be only attraction mode in a pacman collection.
I tested the new attraction mode, splashscreen Pacman/Galaga  was ok, but the games give disturbed video on my Liligo T4.
The old project files function ok for the display.  (noticed the display rotation was changed in the new config file)
Pinned it down to the led routine, pin 16 on Liligo T4 needs to be disabled or changed.
Greets,

**Comment by C64fan** [comment]
Hi Till,
the define MASTER_ATTRACT_GAME_TIMEOUT  works fine for pacman, but not for galaga.
With define MASTER_ATTRACT_GAME_TIMEOUT  95000 , pacman is indeed 95 secondes but galaga only 37 seconds.
I will sent a picture when the mini cabinet is ready, will take some weeks to finish.

**Comment by harbaum** [comment]
The game timeout part is not finished yet. This needs some major changes as the current solution of resetting the ESP32 in order to return to the main menu causes some crackling noise which I'd like to avoid if the system is supposed to be sitting quietly in master attract mode without bothering anyone.

**Comment by C64fan** [comment]
Hi Till,

interessting topic about powering the Esp32 devices, there is a Gpo to Enable the 5 volt rail and lipo.

To clarify the different power options :

power via USB without battery
power via USB with a battery : the battery will get charged, and will be used when USB is disconnected, 
power via the +5V. It won't charge the battery. 
This should be done without battery only : the 5V pin is connected directly to the battery (via Q5) as soon as USB is plugged in.
power via the VBUS (which is the 5V pin of USB). 
This is the best approach, as this will charge the battery, and if no power is applied on VBUS then the battery will power the board. 
As can be seen on the schematic, this is the only way to charge the battery, the +5V or +3.3V pins won't do it. 
The problem is that VBUS isn't directly accessible on a pin. 
Of course VBUS is the same trace as the USB power, so ideally this should be used without the USB, even if it would probably work.

I have to Enable for 5 Volt rail use,
 
void setup()
{
pinMode(15, OUTPUT); // to boot with battery...
digitalWrite(15,1);  // and/or power from 5v rail instead of USB
...
}

Greets

**Comment by C64fan** [comment]
Hi Till,

still following the progress, woodwork is done, looking for the right micro joystick on the web.

Can you make a topic about the things must be done to patch the roms and the pitfalls where to patch, to embed more roms ?
( start address / file format )
"
A set of python scripts is being used to convert and patch the ROM data and emulator code
"
Greets,

**Comment by harbaum** [comment]
You cannot simply add more ROMs and be done. For any new ROM you must also implement the hardware as well and analyze the graphics and sound ROMs and decide which format you want to convert them into to better suit the capabilities of the ESP32.

**Comment by C64fan** [comment]
Hi,
waited some updates and uploaded latest today with all the roms, all ok.
Something changed on the Master attract mode, the menu is visible but the game attract mode does not start for a random game after booting the esp32.
It stays on the game menu forever, former release went to the attract mode of a random game.
In config.h there is no #define MASTER_ATTRACT_GAME_TIMEOUT   anymore. 
The line in config.h is not clear to me, 
#define MASTER_ATTRACT_MENU_TIMEOUT  2000   // start games randomly while sitting idle in menu for 20 seconds, undefine to disable
if I undefine by  //#define MASTER_ATTRACT_MENU_TIMEOUT  2000 , we get compilation error.
How to activate Master_Attract_mode in latest release ?
Greets

**Reply by harbaum** [reply]
I'll check that asap. Enabling the define should usually be sufficient.

**Comment by harbaum** [comment]
The game timeout option has never been used/finished. It was supposed to return from a game in attract mode after some time. So in attract mode Galagino would run attract modes of the different games in turn.

The option to leave the menu after 20 seconds sitting idle is still there and enabled by default  by option "MASTER_ATTRACT_MENU_TIMEOUT  20000"

**Comment by C64fan** [comment]
Hi Till,

if i understand correctly, you are still working on the attract mode, it is not finished yet.
It did work in earlier version for Pacman and Galaga, it went after boot of esp in a loop routine between both games in attraction mode.
The option MASTER_ATTRACT_GAME_TIMEOUT did work in earlier version.
With the latest release and all the roms enabled it stays at the game menu, no attraction mode is started, which is a pity.
What is the plan for enabling the attract mode for all roms after boot of esp in next version ?

Greets from a fan

**Comment by OSManfred** [comment]
Thank you very much for this cool project. I succeeded in converting the ROMs without any trouble.

In the next step, compiling with Arduino IDE I get this error message: I2S_MODE_DAC_BUILT_IN' was not declared in this scope

Do You have any idea about the reason?

Thank you very much!

In file included from C:\Users\Lappi\Documents\Arduino\libraries\FastLED\src/FastLED.h:75,
from E:\galagino-main\galagino\leds.h:8,
from E:\galagino-main\galagino\leds.cpp:7:
C:\Users\Lappi\Documents\Arduino\libraries\FastLED\src/fastspi.h:157:23: note: #pragma message: No hardware SPI pins defined. All SPI access will default to bitbanged output
# pragma message "No hardware SPI pins defined. All SPI access will default to bitbanged output"
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from C:\Users\Lappi\Documents\Arduino\libraries\FastLED\src/platforms/esp/32/fastled_esp32.h:12,
from C:\Users\Lappi\Documents\Arduino\libraries\FastLED\src/platforms.h:42,
from C:\Users\Lappi\Documents\Arduino\libraries\FastLED\src/FastLED.h:62,
from E:\galagino-main\galagino\leds.h:8,
from E:\galagino-main\galagino\leds.cpp:7:
C:\Users\Lappi\Documents\Arduino\libraries\FastLED\src/platforms/esp/32/clockless_rmt_esp32.h: In instantiation of 'class ClocklessController<16, 40, 100, 60, (EOrder)66, 0, false, 5>':
C:\Users\Lappi\Documents\Arduino\libraries\FastLED\src/chipsets.h:654:7: required from 'class WS2812Controller800Khz<16, (EOrder)66>'
C:\Users\Lappi\Documents\Arduino\libraries\FastLED\src/FastLED.h:119:34: required from 'class NEOPIXEL<16>'
C:\Users\Lappi\Documents\Arduino\libraries\FastLED\src/FastLED.h:366:28: required from 'static CLEDController& CFastLED::addLeds(CRGB*, int, int) [with CHIPSET = NEOPIXEL; unsigned char DATA_PIN = 16]'
E:\galagino-main\galagino\leds.cpp:21:52: required from here
C:\Users\Lappi\Documents\Arduino\libraries\FastLED\src/platforms/esp/32/clockless_rmt_esp32.h:358:46: error: static assertion failed: Invalid pin specified
static_assert(FastPin<DATA_PIN>::validpin(), "Invalid pin specified");
~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
In file included from C:\Users\Lappi\Documents\Arduino\libraries\FastLED\src/FastLED.h:75,
from E:\galagino-main\galagino\leds.h:8,
from E:\galagino-main\galagino\galagino.ino:14:
C:\Users\Lappi\Documents\Arduino\libraries\FastLED\src/fastspi.h:157:23: note: #pragma message: No hardware SPI pins defined. All SPI access will default to bitbanged output
# pragma message "No hardware SPI pins defined. All SPI access will default to bitbanged output"
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
E:\galagino-main\galagino\galagino.ino: In function 'void audio_init()':
E:\galagino-main\galagino\galagino.ino:636:58: error: 'I2S_MODE_DAC_BUILT_IN' was not declared in this scope
.mode = (i2s_mode_t)(I2S_MODE_MASTER | I2S_MODE_TX | I2S_MODE_DAC_BUILT_IN),
^~~~~~~~~~~~~~~~~~~~~
E:\galagino-main\galagino\galagino.ino:636:58: note: suggested alternative: 'I2S_MODE_MASTER'
.mode = (i2s_mode_t)(I2S_MODE_MASTER | I2S_MODE_TX | I2S_MODE_DAC_BUILT_IN),
^~~~~~~~~~~~~~~~~~~~~
I2S_MODE_MASTER
E:\galagino-main\galagino\galagino.ino:665:20: error: 'I2S_DAC_CHANNEL_RIGHT_EN' was not declared in this scope
i2s_set_dac_mode(I2S_DAC_CHANNEL_RIGHT_EN);
^~~~~~~~~~~~~~~~~~~~~~~~
E:\galagino-main\galagino\galagino.ino:665:20: note: suggested alternative: 'I2S_CHANNEL_FMT_RIGHT_LEFT'
i2s_set_dac_mode(I2S_DAC_CHANNEL_RIGHT_EN);
^~~~~~~~~~~~~~~~~~~~~~~~
I2S_CHANNEL_FMT_RIGHT_LEFT
E:\galagino-main\galagino\galagino.ino:665:3: error: 'i2s_set_dac_mode' was not declared in this scope
i2s_set_dac_mode(I2S_DAC_CHANNEL_RIGHT_EN);
^~~~~~~~~~~~~~~~
E:\galagino-main\galagino\galagino.ino:665:3: note: suggested alternative: 'i2s_set_clk'
i2s_set_dac_mode(I2S_DAC_CHANNEL_RIGHT_EN);
^~~~~~~~~~~~~~~~
i2s_set_clk

exit status 1

Compilation error: 'I2S_MODE_DAC_BUILT_IN' was not declared in this scope

**Reply by harbaum** [reply]
What versions of the Arduino IDE and the ESP32 board support did you use? Did you select the esp32 dev board?

**Reply by OSManfred** [reply]
I use Arduino IDE 2.11 and the ESP32 Dev Kit C V4 ESP32-WROOM-32

Now I selevted die esp32 dev board, now compiling works! Thank you! Now I´m wiring the ESP32 with the Display. I´ll keep you up to date.

**Comment by C64fan** [comment]
Got the attraction mode working.
It was a combination of pinning which differs on the TTGo board.
Added a timer of 90 seconds with a emulation_reset(); in the loop function.
Now we get the Main menu and random roms as attraction every 90 seconds in a loop.
Great project and a lot of fun.

**Comment by Ste-Rimini** [comment]
Hi, I need help.
When compiling the sketch galagino the following error appears:
In file included from E:\Arduino\Galagino_2.0\Sketch\ESP32\galagino\galagino.ino:56:
E:\Arduino\Galagino_2.0\Sketch\ESP32\galagino\pacman.h:95:10: fatal error: pacman_logo.h: No such file or directory
  #include "pacman_logo.h"
           ^~~~~~~~~~~~~~~
compilation terminated.

exit status 1

Compilation error: pacman_logo.h: No such file or directory

Can anyone help me?
Thank you

**Comment by harbaum** [comment]
Does the file exist in your Galagino directory? You need to download all necessary files.

**Comment by Ste-Rimini** [comment]
Hi, I compile galagino but in 1942.h the following 5 files are called which I cannot find.
The 5 files are:
#include "1942_rom1.h"
#include "1942_rom2.h"
#include "1942_rom1_b0.h"
#include "1942_rom1_b1.h"
#include "1942_rom1_b2.h"

Where do I find it?
Thank you

**Reply by harbaum** [reply]
None of the ROMs are included for legal reasons. You need to generate these files yourself as described here:

https://github.com/harbaum/galagino/tree/main/roms

**Comment by Ste-Rimini** [comment]
HI,
I converted all the ROMs but an error persists:
Compilation error: 1942_rom1.h: No such file or directory
I successfully started the "sh conv_win.sh" command, it ran without any errors.

How can I solve it?
Thank you

**Reply by harbaum** [reply]
I have never used the windows script myself. According to https://github.com/harbaum/galagino/issues/29 the 1942 entries are missing there. Please copy them from https://github.com/harbaum/galagino/blob/main/romconv/conv.sh and test if they work under Windows as well. 

If they do, then please submit a PR.

**Reply by Ste-Rimini** [reply]
Works well !!!
Thank you very much.

**Reply by harbaum** [reply]
Please send a PR
